### PR TITLE
Destroy dependent entities when parent is deleted

### PR DESCRIPTION
**Pull request by miha-plesko**
Zombie children were remaining in VMDB when DELETE targeted refresh was triggered on parent entity because Nuage doesn't emit events for underlying children (which get implicitly deleted).

With this commit we tell Rails to always remove all the children in VMDB as well to avoid the zombie thing.

### CODE REVIEW

File: app/models/manageiq/providers/nuage/network_manager/cloud_subnet.rb [hunk 1]
```diff
@@ -1,2 +1,7 @@
 class ManageIQ::Providers::Nuage::NetworkManager::CloudSubnet < ::CloudSubnet
+  before_destroy :remove_network_ports, prepend: true
+
+  def remove_network_ports
```
Comment (miha-plesko) [inline app/models/manageiq/providers/nuage/network_manager/cloud_subnet.rb]: @Ladas so this manual deletion is needed because Rails won't remove actual NetworkPorts based on `:dependent => :destroy` but only the records in intermediate table.

Comment (Ladas) [reply app/models/manageiq/providers/nuage/network_manager/cloud_subnet.rb]: hm, we usually delete the network ports as part of the Vms

can you describe a detail usecase/workflow for this to happen?

**Comment by miha-plesko** [comment]
This is what happens when user deletes Enterprise i.e. CloudTenant:

- user deletes CloudTenant on Nuage server
- Nuage server emits event to notify MIQ that CloudTenant was deleted
- but in fact Nuage server removes no only CloudTenant, but also all its children: routers, floating ips, subnets, ports, security groups etc.
- but MIQ only recieved event that CloudTenant was deleted
- by default CloudTenant removal doesn't destroy all children
- so we update logic here to match what happens on Nuage server

It's similar when user removes NetworkRouter on NuageServer: all its children are removed but only one event is emitted.

In other words: we have Ansible running in development environment and it periodically creates some entities and then removes tenant. Result was that VMDB was having more and more CloudSubnets and NetworkPorts with same name. Full EMS refresh removes them, but I would prefer to remove them earlier. Looking forward for your opinion.

File: app/models/manageiq/providers/nuage/network_manager/cloud_tenant.rb [hunk 2]
```diff
@@ -1,4 +1,10 @@
 class ManageIQ::Providers::Nuage::NetworkManager::CloudTenant < ::CloudTenant
+  has_many   :security_groups, :dependent => :destroy
+  has_many   :cloud_subnets,   :dependent => :destroy
+  has_many   :network_ports,   :dependent => :destroy
```
Comment (Ladas) [inline app/models/manageiq/providers/nuage/network_manager/cloud_tenant.rb]: so, this should remove all network_ports, is the case that these are some owned by other tenant?

Comment (Ladas) [reply app/models/manageiq/providers/nuage/network_manager/cloud_tenant.rb]: ok, makes sense. I don't think we have a better way to do this now

Comment (Ladas) [reply app/models/manageiq/providers/nuage/network_manager/cloud_tenant.rb]: btw. nuage doesn't send events for all the objects that are deleted?

Comment (miha-plesko) [reply app/models/manageiq/providers/nuage/network_manager/cloud_tenant.rb]: As far as I've tested it doesn't. It only sends event for explicitly deleted entity while it silently removes its children..

Comment (Ladas) [reply app/models/manageiq/providers/nuage/network_manager/cloud_tenant.rb]: hm strange, in other providers, we usually get event per each entity
